src/ring_theory/integral_domain.lean

```diff
@@ -15,33 +15,45 @@ import algebra.geom_sum
 
 section
 
-open finset polynomial
-
-variables {R : Type*} [integral_domain R] (S : set (units R)) [is_subgroup S] [fintype S]
+open finset polynomial function
+open_locale big_operators nat
 
-lemma card_nth_roots_subgroup_units [decidable_eq R] {n : ℕ} (hn : 0 < n) (a : S) :
-  (univ.filter (λ b : S, b ^ n = a)).card ≤ (nth_roots n ((a : units R) : R)).card :=
-card_le_card_of_inj_on (λ a, ((a : units R) : R))
-  (by simp [mem_nth_roots hn, (units.coe_pow _ _).symm, -units.coe_pow, units.ext_iff.symm, subtype.coe_ext])
-  (by simp [units.ext_iff.symm, subtype.coe_ext.symm])
+variables {R : Type*} {G : Type*} [integral_domain R] [group G] [fintype G]
 
-instance subgroup_units_cyclic : is_cyclic S :=
-by haveI := classical.dec_eq R; exact
-is_cyclic_of_card_pow_eq_one_le
-  (λ n hn, le_trans (card_nth_roots_subgroup_units S hn 1) (card_nth_roots _ _))
+lemma card_nth_roots_subgroup_units (f : G →* R) (hf : injective f) {n : ℕ} (hn : 0 < n) (g₀ : G) :
+  ({g ∈ univ | g ^ n = g₀} : finset G).card ≤ (nth_roots n (f g₀)).card :=
+begin
+  apply card_le_card_of_inj_on f,
+  { intros g hg, rw [sep_def, mem_filter] at hg, rw [mem_nth_roots hn, ← f.map_pow, hg.2] },
+  { intros, apply hf, assumption }
+end
 
+/-- A finite subgroup of the unit group of an integral domain is cyclic. -/
+lemma is_cyclic_of_subgroup_integral_domain (f : G →* R) (hf : injective f) : is_cyclic G :=
+begin
+  haveI := classical.dec_eq G,
+  apply is_cyclic_of_card_pow_eq_one_le,
+  intros n hn,
+  convert (le_trans (card_nth_roots_subgroup_units f hf hn 1) (card_nth_roots n (f 1)))
 end
 
 section
-variables {G : Type*} {R : Type*} [group G] [integral_domain R]
 
-open_locale big_operators nat
+variables (S : set (units R)) [is_subgroup S] [fintype S]
 
-open finset
+/-- A finite subgroup of the units of an integral domain is cyclic. -/
+instance subgroup_units_cyclic : is_cyclic S :=
```

Comment:
I wonder whether this instance is even needed, in light of `is_cyclic_of_subgroup_integral_domain`? It doesn't look like a good candidate to be an instance since the head `S` is a variable. What about just deleting it?

Reply:
It was already there. I'm fine with deleting. @ChrisHughes24 What do you think?

Reply:
Does anything break if you delete it?

Reply:
Yes, it's used on line 90 of that file (two lemmas down):
```
from is_cyclic.exists_monoid_generator (set.range (f.to_hom_units)),
```

Reply:
Hmm, I'm not that much of a fan of this magic action-at-a-distance, but okay.